Lib/test/test_glob.py

```diff
@@ -513,7 +513,14 @@ def fn(pat):
             return glob.translate(pat, recursive=True, include_hidden=True, seps=['/', '\\'])
         self.assertEqual(fn('foo/bar\\baz'), r'(?s:foo[/\\]bar[/\\]baz)\Z')
         self.assertEqual(fn('**/*'), r'(?s:(?:.+[/\\])?[^/\\]+)\Z')
-
+        self.assertEqual(fn('foo[!a]bar'), r'(?s:foo(?![/\\])[^a]bar)\Z')
+        self.assertEqual(fn('foo[%-0]bar'), r'(?s:foo(?![/\\])[%-0]bar)\Z')
+        self.assertEqual(fn('foo[%-0][1-9]bar'), r'(?s:foo(?![/\\])[%-0][1-9]bar)\Z')
+        self.assertEqual(fn('foo[0-%]bar'), r'(?s:foo(?!)bar)\Z')
```

Comment:
`[0-%]` is an empty range that should probably be removed here. Otherwise `(?!)` will never match anything.

```py
assert re.match(r'(?s:foo(?!)bar)\Z', 'foobar') is None
assert re.match(r'(?s:foo(?!)bar)\Z', 'foo/bar') is None
```

OTOH, I couldn't find anything in the specs concerning empty ranges. One could say we want a character falling in this empty range (so we need fooXbar with X in this empty range). Since it's impossible, there will not be anything matching that expression, so we could throw away everything afterwards to optimize the regular expression.

**However**, the current implementation always removed empty ranges. So we should likely do the same here and indicate this in the docs as well.